src/test/ui/const-generics/dyn-supertraits.rs

```diff
@@ -0,0 +1,82 @@
+// run-pass
+// revisions: full min
+
+#![cfg_attr(full, feature(const_generics))]
+#![cfg_attr(full, allow(incomplete_features))]
+#![cfg_attr(min, feature(min_const_generics))]
+
+trait Foo<const N: usize> {
+    fn myfun(&self) -> usize;
+}
+trait Bar<const N: usize> : Foo<N> {}
+trait Baz: Foo<3> {}
+
+struct FooType<const N: usize>;
+struct BarType<const N: usize>;
+struct BazType;
+
+impl<const N: usize> Foo<N> for FooType<N> {
+    fn myfun(&self) -> usize { N }
+}
+impl<const N: usize> Foo<N> for BarType<N> {
+    fn myfun(&self) -> usize { N + 1 }
+}
+impl<const N: usize> Bar<N> for BarType<N> {}
+impl Foo<3> for BazType {
+    fn myfun(&self) -> usize { 999 }
+}
+impl Baz for BazType {}
+
+trait Foz {}
+trait Boz: Foo<3> + Foz {}
+trait Bok<const N: usize>: Foo<N> + Foz {}
+
+struct FozType;
+struct BozType;
+struct BokType<const N: usize>;
+
+impl Foz for FozType {}
+
+impl Foz for BozType {}
+impl Foo<3> for BozType {
+    fn myfun(&self) -> usize { 9999 }
+}
+impl Boz for BozType {}
+
+impl<const N: usize> Foz for BokType<N> {}
+impl<const N: usize> Foo<N> for BokType<N> {
+    fn myfun(&self) -> usize { N + 2 }
+}
+impl<const N: usize> Bok<N> for BokType<N> {}
+
+fn a<const N: usize>(_: &dyn Foo<N>) {}
+fn b(_: &dyn Foo<3>) {}
+fn c<T: Bok<N>, const N: usize>(x: T) { a::<N>(&x); }
+fn d<T: ?Sized + Foo<3>>(_: &T) {}
+fn e(x: &dyn Bar<3>) { d(x); }
```

Comment:
hmm, do you think propagating `myfun` upwards here makes sense?
```suggestion
fn a<const N: usize>(x: &dyn Foo<N>) -> usize { x.myfun() }
fn b(x: &dyn Foo<3>) -> usize { x.myfun() }
fn c<T: Bok<N>, const N: usize>(x: T) -> usize { a::<N>(&x); }
fn d<T: ?Sized + Foo<3>>(x: &T) -> usize { x.myfun() }
fn e(x: &dyn Bar<3>) -> usize { d(x); }
```
otherwise r=me